services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/permissions/HankeKayttajaService.kt

```diff
@@ -281,6 +283,31 @@ class HankeKayttajaService(
             } ?: throw HankeNotFoundException(hankeTunnus)
     }
 
+    @Transactional
+    fun updateContactInfo(hankeTunnus: String, update: ContactUpdate, userId: UUID): HankeKayttaja {
+        hankeRepository
+            .findOneByHankeTunnus(hankeTunnus)
+            ?.let { getKayttajaForHanke(userId, it.id) }
+            ?.let { hankeKayttajaEntity ->
+                // changing name is not allowed if the user is identified (has a permission)
+                if (
+                    hankeKayttajaEntity.permission != null &&
+                        (!update.etunimi.isNullOrEmpty() || !update.sukunimi.isNullOrEmpty())
+                ) {
+                    throw UserAlreadyHasPermissionException(
+                        userId.toString(),
+                        hankeKayttajaEntity.id,
+                        hankeKayttajaEntity.permission!!.id
+                    )
+                }
+                hankeKayttajaEntity.sahkoposti = update.sahkoposti
+                hankeKayttajaEntity.puhelin = update.puhelinnumero
+                update.etunimi?.let { hankeKayttajaEntity.etunimi = it }
+                update.sukunimi?.let { hankeKayttajaEntity.sukunimi = it }
+                return hankeKayttajaEntity.toDomain()
+            } ?: throw HankeNotFoundException(hankeTunnus)
```

Comment:
Since the body has additional nesting and is quite long, it would be clearer here to handle the null case first.

Another alternative would be to create a private method from the content of the later let -block.
```suggestion
        val hankeKayttajaEntity =
            hankeRepository.findOneByHankeTunnus(hankeTunnus)?.let {
                getKayttajaForHanke(userId, it.id)
            } ?: throw HankeNotFoundException(hankeTunnus)

        // changing name is not allowed if the user is identified (has a permission)
        if (
            hankeKayttajaEntity.permission != null &&
                (!update.etunimi.isNullOrEmpty() || !update.sukunimi.isNullOrEmpty())
        ) {
            throw UserAlreadyHasPermissionException(
                userId.toString(),
                hankeKayttajaEntity.id,
                hankeKayttajaEntity.permission!!.id
            )
        }

        hankeKayttajaEntity.sahkoposti = update.sahkoposti
        hankeKayttajaEntity.puhelin = update.puhelinnumero
        update.etunimi?.let { hankeKayttajaEntity.etunimi = it }
        update.sukunimi?.let { hankeKayttajaEntity.sukunimi = it }
        return hankeKayttajaEntity.toDomain()
```

Reply:
I changed the code following your suggestion